xline-client/src/clients/auth.rs

```diff
@@ -41,4 +55,697 @@ impl AuthClient {
+        if request.inner.perm.is_none() {
+            return Err(ClientError::InvalidArgs(String::from(
+                "Permission not given",
+            )));
+        }
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Revokes role permission.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_revoke_permission(
+        &self,
+        request: AuthRoleRevokePermissionRequest,
+    ) -> Result<AuthRoleRevokePermissionResponse> {
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Send request using fast path
+    async fn handle_req_fast_path<Req: Into<RequestWrapper>, Res: From<ResponseWrapper>>(
+        &self,
+        request: Req,
+    ) -> Result<Res> {
+        let propose_id = self.generate_propose_id();
+        let request = RequestWithToken::new_with_token(request.into(), self.token.clone());
+        let cmd = Command::new(vec![], request, propose_id);
+        let cmd_res = self.curp_client.propose(cmd).await?;
+        Ok(cmd_res.decode().into())
+    }
+
+    /// Send request using slow path
+    async fn handle_req_slow_path<Req: Into<RequestWrapper>, Res: From<ResponseWrapper>>(
+        &self,
+        request: Req,
+    ) -> Result<Res> {
+        let propose_id = self.generate_propose_id();
+        let request = RequestWithToken::new_with_token(request.into(), self.token.clone());
+        let cmd = Command::new(vec![], request, propose_id);
+        let (cmd_res, sync_res) = self.curp_client.propose_indexed(cmd).await?;
+        let mut res_wrapper = cmd_res.decode();
+        res_wrapper.update_revision(sync_res.revision());
+        Ok(res_wrapper.into())
+    }
+
+    /// Generate hash of the password
+    fn hash_password(password: &[u8]) -> String {
+        let salt = SaltString::generate(&mut OsRng);
+        #[allow(clippy::panic)] // This doesn't seems to be fallible
+        let hashed_password = Pbkdf2
+            .hash_password(password, salt.as_ref())
+            .unwrap_or_else(|e| panic!("Failed to hash password: {e}"));
+        hashed_password.to_string()
+    }
+
+    /// Generate a new `ProposeId`
+    fn generate_propose_id(&self) -> ProposeId {
+        ProposeId::new(format!("{}-{}", self.name, Uuid::new_v4()))
+    }
+}
+
+/// Request for `Authenticate`
+#[derive(Debug)]
+pub struct AuthenticateRequest {
+    /// inner request
+    inner: xlineapi::AuthenticateRequest,
+}
+
+impl AuthenticateRequest {
+    /// New `AuthenticateRequest`
+    #[inline]
+    pub fn new(name: impl Into<String>, password: impl Into<String>) -> Self {
+        Self {
+            inner: xlineapi::AuthenticateRequest {
+                name: name.into(),
+                password: password.into(),
+            },
+        }
+    }
+}
+
+impl From<AuthenticateRequest> for xlineapi::AuthenticateRequest {
+    #[inline]
+    fn from(req: AuthenticateRequest) -> Self {
+        req.inner
+    }
+}
```

Comment:
I suggest move all wrap types and the conversion to a separate file

Reply:
done